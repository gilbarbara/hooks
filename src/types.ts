import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
