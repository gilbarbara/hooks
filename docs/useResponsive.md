# useResponsive

Get responsive breakpoints.

## Usage

```tsx
import { useResponsive } from '@gilbarbara/hooks';

function Component() {
  const { between, min, max, orientation, size } = useResponsive();

  return (
    <div>
      {max('sm', 'landscape') && <h1>Extra Small</h1>}
      {between('sm', 'lg') && <h1>Between small and large</h1>}
      <p>Extra Small {min('xs') ? '✔' : '✖️'}</p>
      <p>Small {min('sm') ? '✔' : '✖️'}</p>
      <p>Medium {min('md') ? '✔' : '✖️'}</p>
      <p>Large {min('lg') ? '✔' : '✖️'}</p>
      <p>Extra Large {min('xl') ? '✔' : '✖️'}</p>
      <footer>
        {size} - {orientation}
      </footer>
    </div>
  );
}
```

## Reference

```typescript
const defaultBreakpoints = { xs: 0, sm: 400, md: 768, lg: 1024, xl: 1280 };

type UseResponsiveOrientation = 'landscape' | 'portrait';

interface UseResponsiveResult<T> {
  between(min: keyof T, max: keyof T, andOrientation?: UseResponsiveOrientation): boolean;
  min(breakpoint: keyof T, andOrientation?: UseResponsiveOrientation): boolean;
  max(breakpoint: keyof T, andOrientation?: UseResponsiveOrientation): boolean;
  orientation: UseResponsiveOrientation;
  size: keyof T;
}

useResponsive<T extends Record<string, number> | typeof defaultBreakpoints>(
  breakpoints?: T,
  initialWidth = Infinity,
  initialHeight = Infinity
): UseResponsiveResult<T>;
```
