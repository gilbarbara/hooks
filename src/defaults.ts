import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
