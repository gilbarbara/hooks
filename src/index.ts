export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
