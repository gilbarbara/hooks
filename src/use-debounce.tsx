import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
