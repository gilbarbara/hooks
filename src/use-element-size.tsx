import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
