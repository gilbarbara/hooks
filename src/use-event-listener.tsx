import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
