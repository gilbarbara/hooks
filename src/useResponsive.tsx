import { useCallback, useEffect, useState } from 'react';

import { canUseDOM, off, on } from './utils';

const defaultBreakpoints = { xs: 0, sm: 400, md: 768, lg: 1024, xl: 1280 };

export type Breakpoints = typeof defaultBreakpoints;

export type UseResponsiveOrientation = 'landscape' | 'portrait';

export interface UseResponsiveResult<T> {
  between(min: keyof T, max: keyof T, andOrientation?: UseResponsiveOrientation): boolean;
  max(breakpoint: keyof T, andOrientation?: UseResponsiveOrientation): boolean;
  min(breakpoint: keyof T, andOrientation?: UseResponsiveOrientation): boolean;
  orientation: UseResponsiveOrientation;
  size: keyof T;
}

function useResponsiveBase<T extends Record<string, number>>(
  breakpoints: T,
  initialWidth = Infinity,
  initialHeight = Infinity,
): UseResponsiveResult<T> {
  const sizes = Object.entries(breakpoints).sort(([, aSize], [, bSize]) => bSize - aSize);
  const smallestBreakpoint = sizes[sizes.length - 1];

  if (smallestBreakpoint[1] !== 0) {
    if (process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line no-console
      console.warn(`The "${smallestBreakpoint[0]}" breakpoint should be 0`);
    }

    smallestBreakpoint[1] = 0;
  }

  const getScreen = useCallback((): UseResponsiveResult<T> => {
    const height = canUseDOM() ? window.innerHeight : initialHeight;
    const width = canUseDOM() ? window.innerWidth : initialWidth;
    const size = sizes.find(([, s]) => s <= width) || sizes[0];
    const orientation = width > height ? 'landscape' : 'portrait';

    return {
      between(min, max, andOrientation) {
        return (
          width >= breakpoints[min] &&
          width < breakpoints[max] &&
          (!andOrientation || andOrientation === orientation)
        );
      },
      min(breakpoint, andOrientation) {
        return (
          width >= breakpoints[breakpoint] && (!andOrientation || andOrientation === orientation)
        );
      },
      max(breakpoint, andOrientation) {
        return (
          width < breakpoints[breakpoint] && (!andOrientation || andOrientation === orientation)
        );
      },
      orientation,
      size: size[0] as keyof T,
    };
  }, [breakpoints, initialHeight, initialWidth, sizes]);

  const [screen, setScreen] = useState(getScreen());

  useEffect(() => {
    const onResize = () => {
      const current = getScreen();

      if (current.size !== screen.size || current.orientation !== screen.orientation) {
        setScreen(current);
      }
    };

    on(window, 'resize', onResize);

    return () => {
      off(window, 'resize', onResize);
    };
  }, [getScreen, screen, setScreen]);

  return screen;
}

export function useResponsive<T extends Record<string, number> | Breakpoints>(
  breakpoints?: T,
  initialWidth?: number,
  initialHeight?: number,
): UseResponsiveResult<T> {
  return useResponsiveBase(breakpoints || defaultBreakpoints, initialWidth, initialHeight);
}
