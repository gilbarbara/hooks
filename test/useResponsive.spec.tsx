import * as React from 'react';
import { act, render } from '@testing-library/react';

import { Breakpoints, Orientation, useResponsive } from '../src/useResponsive';

declare let window: any;

interface Props {
  breakpoints?: Record<string, number>;
  maximum?: keyof Breakpoints | string;
  minimum?: keyof Breakpoints | string;
  orientation?: Orientation;
}

function Component({ breakpoints, maximum, minimum, orientation }: Props) {
  const { between, max, min, size } = useResponsive(breakpoints);
  let output = !minimum && !maximum;

  if (minimum && maximum) {
    output = between(minimum, maximum, orientation);
  } else if (minimum) {
    output = min(minimum, orientation);
  } else if (maximum) {
    output = max(maximum, orientation);
  }

  if (!output) {
    return null;
  }

  return (
    <div>
      <h1>{size}</h1>
    </div>
  );
}

describe('useResponsive', () => {
  const spy = jest.spyOn(console, 'warn').mockImplementation();

  afterAll(() => {
    spy.mockRestore();
  });

  describe('min', () => {
    it.each([
      [320, 'xs', 'portrait'],
      [320, 'xs', 'landscape'],
      [400, 'sm', 'portrait'],
      [400, 'sm', 'landscape'],
      [768, 'md', 'portrait'],
      [768, 'md', 'landscape'],
      [1024, 'lg', 'portrait'],
      [1024, 'lg', 'landscape'],
      [1280, 'xl', 'portrait'],
      [1280, 'xl', 'landscape'],
    ] as const)('should render properly for %p with %p and %p', (width, min, orientation) => {
      window.innerWidth = width;

      const { container } = render(<Component minimum={min} orientation={orientation} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('max', () => {
    it.each([
      [360, 'sm'],
      [568, 'sm'],
      [568, 'md'],
      [800, 'md'],
      [800, 'lg'],
      [1100, 'lg'],
      [1100, 'xl'],
      [1280, 'xl'],
    ] as const)('should render properly for %p with %p', (width, max) => {
      window.innerWidth = width;

      const { container } = render(<Component maximum={max} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('between', () => {
    it.each([
      [320, 'xs', 'sm'],
      [400, 'sm', 'md'],
      [568, 'sm', 'md'],
      [768, 'md', 'lg'],
      [1024, 'lg', 'xl'],
    ] as const)('should render properly for %p between %s/%s', (width, min, max) => {
      window.innerWidth = width;

      const { container } = render(<Component maximum={max} minimum={min} />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('custom breakpoints', () => {
    const breakpoints = {
      small: 320,
      medium: 700,
      large: 1200,
    };

    it.each([
      [320, 'small', 'medium'],
      [568, 'small', 'medium'],
      [768, 'medium', 'large'],
      [1024, 'medium', 'large'],
      [1280, 'medium', 'large'],
    ])('should render properly for %p between %s/%s', (width, min, max) => {
      window.innerWidth = width;

      const { container } = render(
        <Component breakpoints={breakpoints} maximum={max} minimum={min} />,
      );

      expect(spy).toHaveBeenCalledWith('The "small" breakpoint should be 0');
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('resize', () => {
    beforeAll(() => {
      window.innerWidth = 1024;
    });

    it('should update the matcher with a window.resize', async () => {
      const { container, rerender } = render(<Component />);

      expect(container.firstChild).toMatchSnapshot();

      await act(async () => {
        window.innerWidth = 768;
        window.dispatchEvent(new Event('resize'));
      });

      rerender(<Component />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
