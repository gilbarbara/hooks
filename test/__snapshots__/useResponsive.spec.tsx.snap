// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useResponsive > between > should render properly for %p between 320/xs sm 1`] = `
<div>
  <h1>
    xs
  </h1>
</div>
`;

exports[`useResponsive > between > should render properly for %p between 400/sm md 1`] = `
<div>
  <h1>
    sm
  </h1>
</div>
`;

exports[`useResponsive > between > should render properly for %p between 568/sm md 1`] = `
<div>
  <h1>
    sm
  </h1>
</div>
`;

exports[`useResponsive > between > should render properly for %p between 768/md lg 1`] = `
<div>
  <h1>
    md
  </h1>
</div>
`;

exports[`useResponsive > between > should render properly for %p between 1024/lg xl 1`] = `
<div>
  <h1>
    lg
  </h1>
</div>
`;

exports[`useResponsive > custom breakpoints > should render properly for %p between 320/small medium 1`] = `
<div>
  <h1>
    small
  </h1>
</div>
`;

exports[`useResponsive > custom breakpoints > should render properly for %p between 568/small medium 1`] = `
<div>
  <h1>
    small
  </h1>
</div>
`;

exports[`useResponsive > custom breakpoints > should render properly for %p between 768/medium large 1`] = `
<div>
  <h1>
    medium
  </h1>
</div>
`;

exports[`useResponsive > custom breakpoints > should render properly for %p between 1024/medium large 1`] = `
<div>
  <h1>
    medium
  </h1>
</div>
`;

exports[`useResponsive > custom breakpoints > should render properly for %p between 1280/medium large 1`] = `null`;

exports[`useResponsive > max > should render properly for %p with %p 360 sm 1`] = `
<div>
  <h1>
    xs
  </h1>
</div>
`;

exports[`useResponsive > max > should render properly for %p with %p 568 md 1`] = `
<div>
  <h1>
    sm
  </h1>
</div>
`;

exports[`useResponsive > max > should render properly for %p with %p 568 sm 1`] = `null`;

exports[`useResponsive > max > should render properly for %p with %p 800 lg 1`] = `
<div>
  <h1>
    md
  </h1>
</div>
`;

exports[`useResponsive > max > should render properly for %p with %p 800 md 1`] = `null`;

exports[`useResponsive > max > should render properly for %p with %p 1100 lg 1`] = `null`;

exports[`useResponsive > max > should render properly for %p with %p 1100 xl 1`] = `
<div>
  <h1>
    lg
  </h1>
</div>
`;

exports[`useResponsive > max > should render properly for %p with %p 1280 xl 1`] = `null`;

exports[`useResponsive > min > should render properly for %p with %p and %p 320 xs landscape 1`] = `null`;

exports[`useResponsive > min > should render properly for %p with %p and %p 320 xs portrait 1`] = `
<div>
  <h1>
    xs
  </h1>
</div>
`;

exports[`useResponsive > min > should render properly for %p with %p and %p 400 sm landscape 1`] = `null`;

exports[`useResponsive > min > should render properly for %p with %p and %p 400 sm portrait 1`] = `
<div>
  <h1>
    sm
  </h1>
</div>
`;

exports[`useResponsive > min > should render properly for %p with %p and %p 768 md landscape 1`] = `null`;

exports[`useResponsive > min > should render properly for %p with %p and %p 768 md portrait 1`] = `
<div>
  <h1>
    md
  </h1>
</div>
`;

exports[`useResponsive > min > should render properly for %p with %p and %p 1024 lg landscape 1`] = `
<div>
  <h1>
    lg
  </h1>
</div>
`;

exports[`useResponsive > min > should render properly for %p with %p and %p 1024 lg portrait 1`] = `null`;

exports[`useResponsive > min > should render properly for %p with %p and %p 1280 xl landscape 1`] = `
<div>
  <h1>
    xl
  </h1>
</div>
`;

exports[`useResponsive > min > should render properly for %p with %p and %p 1280 xl portrait 1`] = `null`;

exports[`useResponsive > resize > should update the matcher with a window.resize 1`] = `
<div>
  <h1>
    lg
  </h1>
</div>
`;

exports[`useResponsive > resize > should update the matcher with a window.resize 2`] = `
<div>
  <h1>
    md
  </h1>
</div>
`;
